### DISCUSSION TITLE
How should we get the height and width of raster images (e.g. PNG & WebP) that are imported statically in custom React components?

### DISCUSSION THREAD

**Discussion by tats-u** [facebook/docusaurus#8673] [Q&A] [answered]
Statically imported images in Markdown or MDX are transformed into `<img>` tag with appropriate `width` and `height` values.
However, if you `import` images in MDX and pass them to custom React components, there seems to be no way to get their sizes and pass them to `<img>` unlike Next.js.
Imported images in Next.js contain [not only their paths but also their widths and heights](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/image.tsx#L58-L65), but in Docusaurus we can get only paths of images.
Could someone tell me how to get heights and widths in Docusaurus?

I want to create a component like this and use it from MDX.

```html
<div>
  <img src="screenshot.webp" width="..." height="..." loding="lazy" />
  <div class="arrow" style="top: ...px; left: ...px;" />
  <div class="speech-shape" style="top: ...px; left: ...px;">Additional description on screenshot</div>
  <div class="red-rect" style="top: ...px; left: ...px;" />
  <!-- more overlay shapes for description-->
</div>
```

**Comment by slorber** [comment, marked as answer]
You can use the ideal image plugin, it will add a webpack loader for images that should return more info and generate variants of the image in multiple sizes.
https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-ideal-image
Although it's advised to use it alongside the ideal image component, you can also create your own React component to render the image the way you want.

If this official plugin does not work for you, then you can create your own image plugin, register a Webpack loader and return the image info you need. The ideal image plugin code can be used as an inspiration.

**Reply by tats-u** [reply]
I looked into the plugin. Looks not too bad. I found `lqip-loader` and it need to be fixed.

- Docusaurus and they lack TypeScript support. You can't import raster images without `TypeError`. You have to add multiple `declare module`s in `.d.ts` files. (#8683)
- `lqip-loader` rejects WebP and AVIF images even though its backend `sharp` can handle them. (#8686)
- Docs don't mention about that imported types will be changed once the plugin is activated. It's not mentioned that how to import images and pass them to `<img>` tags in MDX documents will also be changed either. (#8684)
- Imported type after changed is not exported. The type before modified by `lqip-loader` can be found.
